### ISSUE TITLE
bool(ds) should raise a "the truth value of a Dataset is ambiguous" error

### ISSUE THREAD

**Issue by delgadom** [pydata/xarray#6124]
Throwing this out there - happy to be shot down if people are opposed. 

### Current behavior / griping

Currently, coercing a dataset to a boolean invokes [`ds.__bool__`](https://github.com/pydata/xarray/blob/main/xarray/core/dataset.py#L1451-L1452), which in turn calls `bool(ds.data_vars)`:

```python
class Dataset(DataWithCoords, DatasetArithmetic, Mapping):
    ...
    def __bool__(self) -> bool:
        return bool(self.data_vars)
```

This has the unfortunate property of returning True as long as there is at least one data_variable, regardless of the contents.

Currently, the behavior of `Dataset.__bool__` is, at least as far as I've seen, never helpful but frequently unhelpful. I've seen (and written) tests written for DataArrays being passed a Dataset and suddenly the tests are meaningless _so many times_. Conversely, I've never found a legitimate use case for `bool(ds)`. As far as I can tell, this is essentially the same as `len(ds.data_vars) > 0`.

In fact, while testing out my proposed changes below on a fork, I found two tests in the xarray test suite that had succumbed to this issue: see https://github.com/pydata/xarray/pull/6122 and https://github.com/pydata/xarray/pull/6123.

This *has* been discussed before - see https://github.com/pydata/xarray/issues/4290. This discussion focused on the question "should ` bool(xr.Dataset({'a': False}))` return False?". I agree that it's not clear when it should be false and picking a behavior which deviates from Mapping feels arbitrary and gross.

### Proposed behavior

I'm proposing that the API be changed, so that ` bool(xr.Dataset({'a': False}))` raise an error, similar to the implementation in [`pd.Series`](https://github.com/pandas-dev/pandas/blob/v1.3.5/pandas/core/generic.py#L1535-L1542).

In this implementation in pandas, attempting to evaluate even a single-element series as a boolean raises an error:

```python
In [14]: bool(pd.Series([False]))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-14-b0ad7f4d9277> in <module>
----> 1 bool(pd.Series([False]))

~/miniconda3/envs/rhodium-env/lib/python3.9/site-packages/pandas/core/generic.py in __nonzero__(self)
   1532     @final
   1533     def __nonzero__(self):
-> 1534         raise ValueError(
   1535             f"The truth value of a {type(self).__name__} is ambiguous. "
   1536             "Use a.empty, a.bool(), a.item(), a.any() or a.all()."

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
```

I understand hesitancy around changing the core API. That said, if anyone can find an important, correct use of `bool(ds)` in the wild I'll eat my hat :)

### Implementation

This could be as simple as raising an error on `ds.__bool__`, something like:

```python
class Dataset(DataWithCoords, DatasetArithmetic, Mapping):
    ...
    def __bool__(self) -> bool:
        raise ValueError(
            "The truth value of a Dataset is ambiguous. Reduce the data "
            "variables to a scalar value with any(ds.values()) or "
            "all(ds.values())."
        )
```

The only other change that would be needed is an assertion that directly calls `bool(ds)` in [test_dataset::TestDataset.test_properties](https://github.com/pydata/xarray/blob/main/xarray/tests/test_dataset.py#L547), which checks for the exact behavior I'm changing:

```python
        assert bool(ds)
```

This would need to be changed to:

```python
        with pytest.raises(ValueError):
            bool(ds)
```

If this sounds good, I can submit a PR with these changes.

**Comment by Illviljan** [comment]
A `Dataset` is more similar to a `dict` or `pd.DataFrame`.

DataFrame has a similar error, same cooks I suppose:
```python
bool(pd.DataFrame())
*** ValueError: The truth value of a DataFrame is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

bool(pd.DataFrame([[0, 2], [0, 4]], columns=['A', 'B']))
*** ValueError: The truth value of a DataFrame is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
```

`dict` works the same way as a dataset, if something exist in it is `True`:
```python
bool({})
False

bool({'a': False})
True
```
I see "if not empty do x"-checks all the time with dicts in python code, 
Is it that strange to follow the behavior of `dict`?

**Comment by max-sixty** [comment]
I definitely empathize with the tradeoff here. That you found xarray's test's were making this error is fairly damning.

But the biggest impediment to changing this behavior is that `Dataset` follows the `Mapping` protocol, which has this behavior. One nice feature of xarray is that we follow python's protocols where possible, and that includes truthiness for dict-like / `Mapping` objects. Notably `pd.DataFrame` objects only partially implement the protocol, including `.values`.

If there's a synthesis of keeping the truthiness while reducing the chance of these mistakes, that would be very welcome.

I'm not sure this is an improvement, but in the example converting to a `DataArray` gets away from the truthiness issue: `(result.min(...) >= 1.5).to_array().all()`

(I wrote this before seeing @Illviljan 's response, which is very similar)

**Comment by delgadom** [comment]
Yeah… I do understand how it’s currently working and why, and the behavior is certainly intuitive to those who appreciate the mapping inheritance.

That said, I feel I have to make a last stand argument because this trips people up quite often (on my team and elsewhere). I haven’t yet come across an example of anyone using this correctly, but I see users misusing it all the time.

The examples and behavior you’re showing @Illviljan seem to me like more the natural result of an implementation detail than a critical principle of the dataset design. While it’s obvious why `bool(ds)` resolves to True or False in the current implementation, I still think it’s not used and is out of step with how most users think of a dataset. The fact that nothing in the package requires a change in order to pass tests (the broken tests and the pathological case aside) reaffirms for me that this is not a useful feature.

I don’t know much about the mapping protocol or how closely it must be followed. Is the idea here that packages building on xarray (or interoperability features in e.g. numpy or dask) depend on a strict adherence to the full spec?

**Comment by delgadom** [comment]
> If there's a synthesis of keeping the truthiness while reducing the chance of these mistakes, that would be very welcome.

@max-sixty im not sure what this would look like. Do you mean a warning or are you hinting that the bar that would need to be met is a silver bullet that preserves bool(ds) but somehow isn’t confusing?

**Comment by shoyer** [comment]
The original intention here was definitely to honor the `Mapping` protocol exactly, but I agree that `bool()` is hard to use correctly on `Dataset` objects. I would support deprecating this.

**Comment by delgadom** [comment]
I realize this may be a larger discussion, but the implementation is so easy I went ahead and filed a PR that issues a PendingDeprecationWarning in `Dataset.__bool__`.

**Comment by max-sixty** [comment]
Would this mean that `Dataset` doesn't inherit from `Mapping`? Or that we inherit from `Mapping` but override `bool` to raise an error?

**Comment by shoyer** [comment]
@dcherian So I guess we decided to go for it? :)

@max-sixty We can still inherit from Mapping and just explicitly raise `TypeError` inside `Dataset.__bool__` rather than using the default implementation. This is certainly a slight violation from the Mapping protocol, but `xarray.Dataset` is already a little inconsistent (e.g., we override `__eq__` inconsistently with the default mapping behavior). Overall I think it's probably a worthwhile the more user-friendly choice?

**Comment by dcherian** [comment]
Sorry! I thought we had consensus but perhaps not?

Shall we revert? That said it's `PendingDeprecationWarning` so it should only be raised in tests, which seems OK?

**Comment by max-sixty** [comment]
TBC, I am fine merging! I would lightly vote against it, but weigh my vote far below @shoyer 's.

And more broadly let's not wait for everyone to agree on everything!

**Comment by Illviljan** [comment]
I do wonder at what point a mapping isn't a mapping anymore? 

For example DataFrames aren't considered mappings:
```python
isinstance(df, collections.abc.Mapping)
Out[4]: False
```
And if we are to follow pandas example maybe we should just remove the Mapping inheritance?
https://github.com/pydata/xarray/blob/60754fdbc4ecd9eb3c0978e82635c6d43e8d485b/xarray/core/dataset.py#L584

**Comment by shoyer** [comment]
After discussing this a little more, I am on the fence about whether this change this is a good idea. If we can't come to concensus about expected behavior, then that is probably an indication that we should leave things the same to avoid churn.

**Comment by shoyer** [comment]
I made a Twitter poll, let's see what that says 😄 

https://twitter.com/xarray_dev/status/1478776987925684224?s=20

**Comment by dopplershift** [comment]
$0.02 from the peanut gallery is that my mental model of `Dataset` is that it's a dictionary on steroids, so following the `Mapping` protocol (including `__bool__`) makes sense to me. I put "I don't know" for @shoyer's poll, but if the question was "should" then I would have said "number of variables".

While I'm not going to sit here and argue that `if ds:` is a common operation and important feature (I don't often open files that result in empty datasets), I'd personally want a truly compelling argument for breaking from the `Mapping` protocol. Right now it's "`Dataset` works like `dict` *plus* some stuff". It would change to be "`Dataset` works like `dict` for all things except `__bool__`, plus some stuff". The latter to me requires more mental effort since I'm no longer strictly supplementing past experience.

**Comment by jhamman** [comment]
I'm also late to the party but I would say I fall squarely in the Dataset is a dict-like camp. If we remove `__bool__`, should we also remove `__len__`? Basically, everything @dopplershift aligns with my perspective here.

**Comment by VeckoTheGecko** [comment]
Heyall, another side-effect of this is that

```py
ds in ("some", "collection")
```
returns `True` (which of course, doesn't make sense since an xarray dataset is not contained in that tuple of strings). This is quite annoying if you want to simply write guard clauses like

```python
def some_calculation(ds: xr.Dataset, calculation_type: Literal["polar", "cartesian"]) -> xr.Dataset:
    if calculation_type not in ("polar", "cartesian"):
        raise ValueError(f"Invalid calculation_type. Expected 'polar' or 'cartesian'. Got {type=!r}")
    ...
    return ds
```
since now passing `calculation_type=ds` causes the function to magically continue, and you have to code in some additional logic.

See #11209 for full discussion about this particular problem (which is now closed as duplicate pointing to this thread).

**Comment by dcherian** [comment]
@VeckoTheGecko Can we handle that case with `__contains__`?

**Comment by jsignell** [comment]
> Can we handle that case with `__contains__`?

I thought that as well at first but the `__contains__` that gets called is `tuple.__contains__` (full summary of the codepath in https://github.com/pydata/xarray/issues/11209#issuecomment-3992962366)

**Comment by keewis** [comment]
It looks like we're stuck, then: we do support `ds == scalar`, and `__bool__` behaves like it does for a mapping.

To raise on `ds in sequence`, we'd have to raise in one of them, but I don't think that change is going to be easy to make.

**Comment by jsignell** [comment]
I totally agree that `ds ==scalar` is important and used, we shouldn't change the behavior of that. 

But from the older comments on this issue it doesn't really seem like anyone _uses_ `bool(ds)` so deprecating `__bool__` certainly seems like the less painful option.

**Comment by shoyer** [comment]
I'm pretty sure people write code like `if dataset` to check for non-empty datasets, so I think this would be at least a little painful to change.

**Comment by keewis** [comment]
even though it is used, `ds == scalar` is a bit weird in that different dtypes will behave differently (string == int will always return False, while datetime64 == int may actually compare values).

We may want to actually consider deprecating either that, or `xr.Dataset.__bool__` in favor of `bool(ds.data_vars)` (or both?)

(Depends on how many breaking changes we want to expose our users to, or rather, how much we believe think the cleanup is worth)
